Convex_hull_3/include/CGAL/convex_hull_3.h

```diff
@@ -857,45 +963,78 @@ template <class InputIterator, class Polyhedron_3, class Traits>
 void convex_hull_3(InputIterator first, InputIterator beyond,
                    Polyhedron_3& polyhedron,  const Traits& traits)
 {
-  typedef typename Traits::Point_3                Point_3;  
+  typedef typename Traits::Point_3                Point_3;
   typedef std::list<Point_3>                      Point_3_list;
   typedef typename Point_3_list::iterator         P3_iterator;
+  typedef std::pair<P3_iterator,P3_iterator>      P3_iterator_pair;
+
+  if(first == beyond)
+    return;
 
   Point_3_list points(first, beyond);
-  CGAL_ch_precondition(points.size() > 3);
+  clear(polyhedron);
+
+  if(points.size() == 1)
+  {
+    add_isolated_points(points, polyhedron);
+    return;
+  }
 
-  // at least 4 points 
   typename Traits::Collinear_3 collinear = traits.collinear_3_object();
   typename Traits::Equal_3 equal = traits.equal_3_object();
-
   P3_iterator point1_it = points.begin();
   P3_iterator point2_it = points.begin();
   point2_it++;
 
-  // find three that are not collinear
   while (point2_it != points.end() && equal(*point1_it,*point2_it))
     ++point2_it;
 
-  CGAL_ch_precondition_msg(point2_it != points.end(), 
-        "All points are equal; cannot construct polyhedron.");
-  
+  // if all points are equal, return empty set.
+  if(point2_it == points.end())
+    return;
+
+  if(points.size() == 2)
+  {
+    add_isolated_points(points, polyhedron);
+    return;
+  }
+
   P3_iterator point3_it = point2_it;
   ++point3_it;
-  
-  CGAL_ch_precondition_msg(point3_it != points.end(), 
-        "Only two points with different coordinates; cannot construct polyhedron.");
-  
-  while (point3_it != points.end() && collinear(*point1_it,*point2_it,*point3_it))
-    ++point3_it;
-  
-  CGAL_ch_precondition_msg(point3_it != points.end(), 
-        "All points are collinear; cannot construct polyhedron.");
-  
-  clear(polyhedron);
-  // result will be a polyhedron
-  internal::Convex_hull_3::ch_quickhull_polyhedron_3(points, point1_it, point2_it, point3_it,
-                                                     polyhedron, traits);
 
+  // if only two points have different coordinates then return those.
+  if(point3_it == points.end())
+  {
+    std::list<Point_3> endpoints;
+    find_endpoints(points, endpoints, traits);
+    add_isolated_points(endpoints, polyhedron);
+    return;
+  }
+
+  // if there are 3 points not equal, then the third is not at points.end()
+  if(points.size() == 3)
+  {
+    if(collinear(*point1_it,*point2_it,*point3_it))
+    {
+      std::list<Point_3> endpoints;
+      find_endpoints(points, endpoints, traits);
+      add_isolated_points(endpoints, polyhedron);
```

Comment:
here simply take the 3 points of the 3 iterators and the extreme one are the one that are le larget and the smaller one using a lexicographic sort (something like 2 if and swap should make the job)